File: gamePaths.py
"""Data-driven content paths layered over the shared run systems.

Paths own maps, encounter identity, projectile identity, and boss rosters.  New
path-exclusive encounter builders can be registered here without changing the
leveling, statistics, HUD, or core game loop.
"""

from dataclasses import dataclass
from math import pi
import random

import pygame

import background as bG
from enemyTypes import ENEMY_CATALOG


@dataclass(frozen=True)
class EnemyStyle:
    speed: float = 1.0
    size: float = 1.0
    health: float = 1.0
    damage: float = 1.0
    attack_cooldown: float = 1.0
    attack_range: float = 1.0
    awareness: float = 1.0
    experience: float = 1.0
    projectile_speed: float = 1.0
    projectile_size: float = 1.0
    projectile_damage: float = 1.0
    projectile_range: float = 1.0
    projectile_lifetime: float | None = None
    aimed_chance: float = 1.0
    late_health: float = 1.0
    late_damage: float = 1.0
    late_attack_rate: float = 1.0
    late_projectile_speed: float = 1.0
    late_projectile_damage: float = 1.0
    colors: tuple = ()
    tags: tuple = ()


@dataclass(frozen=True)
class GamePath:
    key: str
    title: str
    subtitle: str
    description: str
    mid_boss: str
    final_boss: str
    accent: pygame.Color
    style: EnemyStyle


PATHS = {
    "sound": GamePath(
        "sound", "PATH OF SOUND", "THE DISSONANCE",
        "The original arena: fast patterns, echoes, and open ground.",
        "beaudis", "dissonance", pygame.Color(207, 191, 151), EnemyStyle(),
    ),
    "touch": GamePath(
        "touch", "PATH OF TOUCH", "THE WEIGHT BELOW",
        "A cramped prison sewer of heavy bodies and slow, punishing shots.",
        "bair", "rot", pygame.Color(91, 132, 74), EnemyStyle(
            speed=.72, size=1.18, health=1.45, damage=1.20,
            attack_cooldown=1.35, awareness=.92, experience=1.18,
            projectile_speed=.68, projectile_size=1.24, projectile_damage=1.28,
            late_health=1.18, late_damage=1.12, late_attack_rate=.68,
            late_projectile_speed=1.08, late_projectile_damage=1.06,
            colors=(pygame.Color(79, 101, 55), pygame.Color(103, 91, 55),
                    pygame.Color(54, 83, 55), pygame.Color(116, 105, 63)),
            tags=("rotton", "heavy"),
        ),
    ),
    "sight": GamePath(
        "sight", "PATH OF SIGHT", "THE QUICKENED HORIZON",
        "An exposed field of small, fragile hunters and close, rapid attacks.",
        "ishe", "chronos", pygame.Color(104, 190, 222), EnemyStyle(
            speed=1.38, size=.78, health=.62, damage=.82,
            attack_cooldown=.68, attack_range=.60, awareness=1.08,
            experience=.82, projectile_speed=1.55, projectile_size=.72,
            projectile_damage=.76, projectile_range=.48,
            late_health=1.12, late_damage=1.20, late_attack_rate=.72,
            late_projectile_speed=1.08, late_projectile_damage=1.12,
            colors=(pygame.Color(105, 190, 220), pygame.Color(135, 210, 230),
                    pygame.Color(228, 142, 63), pygame.Color(244, 174, 82)),
            tags=("sighted", "quick", "close_range"),
        ),
    ),
    "chemesthesis": GamePath(
        "chemesthesis", "PATH OF CHEMESTHESIS", "THE BURNING FIELD",
        "Durable carriers seed long-lived, mostly unaimed hazards everywhere.",
        "kage", "ache", pygame.Color(207, 83, 45), EnemyStyle(
            speed=.94, size=1.05, health=1.55, damage=.92,
            attack_cooldown=1.12, awareness=1.05, experience=1.25,
            projectile_speed=.64, projectile_size=1.08, projectile_damage=.88,
            projectile_range=2.6, projectile_lifetime=12.0, aimed_chance=.22,
            late_health=1.15, late_damage=1.18, late_attack_rate=.62,
            late_projectile_speed=1.10, late_projectile_damage=1.12,
            colors=(pygame.Color(171, 62, 36), pygame.Color(211, 91, 38),
                    pygame.Color(92, 120, 50), pygame.Color(126, 48, 39)),
            tags=("chemesthetic", "minefield", "durable"),
        ),
    ),
    "phantasia": GamePath(
        "phantasia", "PATH OF PHANTASIA", "THE ORNATE DREAM",
        "Broad dream courts surround a few extravagant, feature-heavy ruins.",
        "hypno", "malady", pygame.Color(190, 83, 175), EnemyStyle(
            speed=1.02, size=1.03, health=1.12, damage=1.02,
            attack_cooldown=.96, experience=1.08,
            projectile_size=1.08, projectile_range=1.15,
            late_health=1.20, late_damage=1.18, late_attack_rate=.65,
            late_projectile_speed=1.12, late_projectile_damage=1.10,
            colors=(pygame.Color(117, 48, 121), pygame.Color(161, 57, 147),
                    pygame.Color(202, 85, 174), pygame.Color(91, 48, 119)),
            tags=("phantasian", "ornate"),
        ),
    ),
}

selected_key = "sound"
active_key = "sound"
_projectile_rng = random.Random(9071)


def selected():
    return PATHS[selected_key]


def active():
    return PATHS[active_key]


def select(key):
    global selected_key
    if key not in PATHS:
        raise KeyError(f"Unknown game path: {key}")
    selected_key = key


def cycle(direction):
    keys = tuple(PATHS)
    select(keys[(keys.index(selected_key) + direction) % len(keys)])


def activate_selected():
    global active_key
    active_key = selected_key
    bG.configure_battleground(active_key)


def boss_key(midpoint):
    path = active()
    return path.mid_boss if midpoint else path.final_boss


def is_touch():
    return active_key == "touch"


def _late_pressure(level):
    """Return the non-boss escalation from the midpoint through level 19."""
    return max(0.0, min(1.0, (float(level or 0) - 10.0) / 9.0))


def _toward_one(multiplier, pressure):
    return 1.0 + (multiplier - 1.0) * pressure


def apply_enemy_identity(enemy, level=0):
    """Apply the active path's shared identity exactly once to an enemy tree."""
    path = active()
    style = path.style
    if enemy is None or path.key == "sound" or getattr(enemy, "contentPath", None) == path.key:
        return enemy
    enemy.contentPath = path.key
    pressure = _late_pressure(level)
    enemy.speed *= style.speed
    enemy.size *= style.size
    enemy.maxHp = round(enemy.maxHp * style.health
                        * _toward_one(style.late_health, pressure))
    enemy.hp = enemy.maxHp
    enemy.damage = round(enemy.damage * style.damage
                         * _toward_one(style.late_damage, pressure))
    enemy.expValue *= style.experience
    if style.colors:
        family = getattr(enemy, "family", path.key)
        enemy.color = style.colors[sum(map(ord, family)) % len(style.colors)]
    if hasattr(enemy, "attackCooldownMax"):
        enemy.attackCooldownMax *= (style.attack_cooldown
                                    * _toward_one(style.late_attack_rate, pressure))
    if hasattr(enemy, "attackCooldown"):
        enemy.attackCooldown *= (style.attack_cooldown
                                 * _toward_one(style.late_attack_rate, pressure))
    if hasattr(enemy, "attackRangeTiles"):
        enemy.attackRangeTiles *= style.attack_range
    enemy.awarenessRange *= style.awareness
    enemy.disengageRange *= style.awareness
    enemy.interactionTags = frozenset(
        set(getattr(enemy, "interactionTags", ())) | set(style.tags))
    for child in getattr(enemy, "spawnedEnemies", ()):
        apply_enemy_identity(child, level)
    return enemy


# Optional path-only encounter factories receive the same arguments as
# EnemyCatalog.spawn_patrol and return (encounter, enemies) or None.  Keeping the
# registry path-scoped prevents experimental enemies from leaking into Sound.
_EXCLUSIVE_ENCOUNTER_FACTORIES = {key: [] for key in PATHS}


def register_exclusive_encounter(path_key, factory):
    if path_key not in PATHS:
        raise KeyError(f"Unknown game path: {path_key}")
    _EXCLUSIVE_ENCOUNTER_FACTORIES[path_key].append(factory)


class _PathEnemyCatalog:
    """Adapter around the mature shared catalog with path-only extension hooks."""

    @property
    def definitions(self):
        return ENEMY_CATALOG.definitions

    def spawn(self, *args, **kwargs):
        level = args[0] if args else kwargs.get("level", 0)
        return apply_enemy_identity(ENEMY_CATALOG.spawn(*args, **kwargs), level)

    def spawn_encounter(self, *args, **kwargs):
        level = args[0] if args else kwargs.get("level", 0)
        return self._apply_group(ENEMY_CATALOG.spawn_encounter(*args, **kwargs), level)

    def spawn_patrol(self, *args, **kwargs):
        level = args[0] if args else kwargs.get("level", 0)
        for factory in _EXCLUSIVE_ENCOUNTER_FACTORIES[active_key]:
            result = factory(*args, **kwargs)
            if result:
                return self._apply_group(result, level)
        return self._apply_group(ENEMY_CATALOG.spawn_patrol(*args, **kwargs), level)

    @staticmethod
    def _apply_group(result, level=0):
        if result:
            for enemy in result[1]:
                apply_enemy_identity(enemy, level)
        return result


ENCOUNTERS = _PathEnemyCatalog()


def tune_new_projectiles(projectiles, start_index, level=0, late_pressure=True):
    """Apply active path projectile rules to newly emitted hostile shots."""
    path = active()
    style = path.style
    if path.key == "sound":
        return
    pressure = _late_pressure(level) if late_pressure else 0.0
    for projectile in projectiles[start_index:]:
        if getattr(projectile, "contentPath", None) == path.key:
            continue
        projectile.contentPath = path.key
        projectile.speed *= (style.projectile_speed
                             * _toward_one(style.late_projectile_speed, pressure))
        projectile.size *= style.projectile_size
        projectile.damage = round(
            projectile.damage * style.projectile_damage
            * _toward_one(style.late_projectile_damage, pressure))
        projectile.remainingRange *= style.projectile_range
        if style.projectile_lifetime is not None:
            projectile.lifetime = max(projectile.lifetime or 0,
                                      style.projectile_lifetime)
        if style.aimed_chance < 1 and _projectile_rng.random() > style.aimed_chance:
            # Chemesthesis shots preserve their source but seed the room rather
            # than tracking the player, producing a persistent minefield.
            projectile.direction += _projectile_rng.uniform(-pi, pi)
        if style.colors:
            projectile.color = style.colors[-1]


def tune_encounter_caps(caps, level):
    """Turn late non-Sound rooms into bounded horde encounters."""
    result = dict(caps)
    pressure = _late_pressure(level) if active_key != "sound" else 0.0
    result["enemy_cap"] += round(8 * pressure)
    result["threat_cap"] *= 1.0 + .18 * pressure
    result["population_threat_cap"] *= 1.0 + .15 * pressure
    return result


def tune_encounter_pacing(pacing, level):
    result = dict(pacing)
    pressure = _late_pressure(level) if active_key != "sound" else 0.0
    result["patrol_size"] += round(pressure)
    result["max_world_encounters"] += round(pressure)
    result["spawn_interval_seconds"] *= 1.0 - .22 * pressure
    result["curated_chance"] = min(.65, result["curated_chance"] + .12 * pressure)
    return result


def projectile_cap(level, boss_active=False):
    """Bound hostile draw/update cost while preserving a late-run bullet hell."""
    if boss_active:
        return 150
    base = {"sound": 280, "touch": 280, "sight": 380,
            "chemesthesis": 340, "phantasia": 340}.get(active_key, 300)
    return base + round(40 * _late_pressure(level))

File: RotBoiRemastered/World/GamePaths.cs
using Microsoft.Xna.Framework;
using RotBoiRemastered.Entities;

namespace RotBoiRemastered.World;

/// <summary>
/// Shared per-enemy stat multipliers for one content path. Ported from
/// gamePaths.py's EnemyStyle dataclass. Colors/Tags are nullable rather than
/// defaulting to an empty array (C# doesn't allow non-constant default
/// parameter values) -- null means "no override," matching how the Python
/// original's `if style.colors:` truthiness check treated an empty tuple.
/// </summary>
public sealed record EnemyStyle(
    double Speed = 1.0,
    double Size = 1.0,
    double Health = 1.0,
    double Damage = 1.0,
    double AttackCooldown = 1.0,
    double AttackRange = 1.0,
    double Awareness = 1.0,
    double Experience = 1.0,
    double ProjectileSpeed = 1.0,
    double ProjectileSize = 1.0,
    double ProjectileDamage = 1.0,
    double ProjectileRange = 1.0,
    double? ProjectileLifetime = null,
    double AimedChance = 1.0,
    IReadOnlyList<Color>? Colors = null,
    IReadOnlyList<string>? Tags = null);

/// <summary>Ported from gamePaths.py's GamePath dataclass.</summary>
public sealed record GamePath(
    string Key, string Title, string Subtitle, string Description,
    string MidBoss, string FinalBoss, Color Accent, EnemyStyle Style);

/// <summary>
/// Data-driven content paths layered over the shared run systems. Selection,
/// battleground generation, boss rosters, enemy identity, and hostile
/// projectile tuning are wired into GameSession. The optional Python-only
/// exclusive-encounter extension registry has no registered consumers.
/// </summary>
public static class GamePaths
{
    private static readonly Random ProjectileRng = new(9071);
    // Order matters here (matches Python 3.7+ dict insertion order, which
    // gamePaths.py relies on for e.g. title-screen path cycling) -- Paths is
    // the ordered source of truth; PathsByKey is only for O(1) lookup.
    public static readonly IReadOnlyList<GamePath> Paths = new[]
    {
        new GamePath("sound", "PATH OF SOUND", "THE DISSONANCE",
            "The original arena: fast patterns, echoes, and open ground.",
            "beaudis", "dissonance", new Color(207, 191, 151), new EnemyStyle()),
        new GamePath("touch", "PATH OF TOUCH", "THE WEIGHT BELOW",
            "A cramped prison sewer of heavy bodies and slow, punishing shots.",
            "bair", "rot", new Color(91, 132, 74), new EnemyStyle(
                Speed: .70, Size: 1.22, Health: 1.65, Damage: 1.28,
                AttackCooldown: 1.48, Awareness: .92, Experience: 1.22,
                ProjectileSpeed: .68, ProjectileSize: 1.24, ProjectileDamage: 1.28,
                Colors: new[] { new Color(79, 101, 55), new Color(103, 91, 55), new Color(54, 83, 55), new Color(116, 105, 63) },
                Tags: new[] { "rotton", "heavy" })),
        new GamePath("sight", "PATH OF SIGHT", "THE QUICKENED HORIZON",
            "An exposed field of small, fragile hunters and close, rapid attacks.",
            "ishe", "chronos", new Color(104, 190, 222), new EnemyStyle(
                Speed: 1.45, Size: .76, Health: .56, Damage: .78,
                AttackCooldown: .58, AttackRange: .58, Awareness: 1.08,
                Experience: .82, ProjectileSpeed: 1.55, ProjectileSize: .72,
                ProjectileDamage: .76, ProjectileRange: .48,
                Colors: new[] { new Color(105, 190, 220), new Color(135, 210, 230), new Color(228, 142, 63), new Color(244, 174, 82) },
                Tags: new[] { "sighted", "quick", "close_range" })),
        new GamePath("chemesthesis", "PATH OF CHEMESTHESIS", "THE BURNING FIELD",
            "Durable carriers seed long-lived, mostly unaimed hazards everywhere.",
            "kage", "ache", new Color(207, 83, 45), new EnemyStyle(
                Speed: .92, Size: 1.06, Health: 2.15, Damage: 1.02,
                AttackCooldown: .94, Awareness: 1.05, Experience: 1.34,
                ProjectileSpeed: .62, ProjectileSize: 1.08, ProjectileDamage: .88,
                ProjectileRange: 4.0, ProjectileLifetime: 18.0, AimedChance: .18,
                Colors: new[] { new Color(171, 62, 36), new Color(211, 91, 38), new Color(92, 120, 50), new Color(126, 48, 39) },
                Tags: new[] { "chemesthetic", "minefield", "durable" })),
        new GamePath("phantasia", "PATH OF PHANTASIA", "THE ORNATE DREAM",
            "Broad dream courts surround a few extravagant, feature-heavy ruins.",
            "hypno", "malady", new Color(190, 83, 175), new EnemyStyle(
                Speed: 1.02, Size: 1.03, Health: 1.12, Damage: 1.02,
                AttackCooldown: .96, Experience: 1.08,
                ProjectileSize: 1.08, ProjectileRange: 1.15,
                Colors: new[] { new Color(117, 48, 121), new Color(161, 57, 147), new Color(202, 85, 174), new Color(91, 48, 119) },
                Tags: new[] { "phantasian", "ornate" })),
    };

    public static readonly IReadOnlyDictionary<string, GamePath> PathsByKey =
        Paths.ToDictionary(path => path.Key);

    // Module-level mutable state, same as Python's selected_key/active_key.
    // No test class other than GamePathsTests touches this yet -- if one
    // ever does, it'll need the same [Collection("...")] treatment as
    // GameProfileTests/KeybindsTests (see GameProfileStateCollection.cs).
    private static string _selectedKey = "sound";
    private static string _activeKey = "sound";

    public static GamePath Selected() => PathsByKey[_selectedKey];

    public static GamePath Active() => PathsByKey[_activeKey];

    public static void Select(string key)
    {
        if (!PathsByKey.ContainsKey(key))
            throw new KeyNotFoundException($"Unknown game path: {key}");
        _selectedKey = key;
    }

    public static void Cycle(int direction)
    {
        var keys = Paths.Select(path => path.Key).ToList();
        int index = keys.IndexOf(_selectedKey);
        int nextIndex = ((index + direction) % keys.Count + keys.Count) % keys.Count;
        Select(keys[nextIndex]);
    }

    /// <summary>
    /// Activates the selected path and generates its battleground. Unlike
    /// Python's activate_selected() (which reassigned background.py's
    /// module-level currRoomRects etc. in place), this returns the new
    /// Battleground -- there's no hidden global "current battleground" to
    /// mutate here, so the caller is responsible for holding onto it. That
    /// owner will most likely be whatever session/world container gets
    /// built when Entities/ and Core/RotBoiGame.cs's state machine are wired
    /// together.
    /// </summary>
    public static Battleground ActivateSelected()
    {
        _activeKey = _selectedKey;
        return Battleground.CreateForPath(_activeKey);
    }

    public static string BossKey(bool midpoint) => midpoint ? Active().MidBoss : Active().FinalBoss;

    public static bool IsTouch() => _activeKey == "touch";

    /// <summary>Applies the active path's shared enemy identity exactly once.</summary>
    public static Enemy? ApplyEnemyIdentity(Enemy? enemy)
    {
        var path = Active();
        var style = path.Style;
        if (enemy is null || path.Key == "sound" || enemy.ContentPath == path.Key)
            return enemy;

        enemy.ContentPath = path.Key;
        enemy.Speed *= (float)style.Speed;
        enemy.Size *= (float)style.Size;
        enemy.MaxHp = (int)Math.Round(enemy.MaxHp * style.Health);
        enemy.Hp = enemy.MaxHp;
        enemy.Damage = (int)Math.Round(enemy.Damage * style.Damage);
        enemy.ExpValue *= style.Experience;
        if (style.Colors is { Count: > 0 })
            enemy.Color = style.Colors[enemy.Family.Sum(character => character) % style.Colors.Count];
        if (enemy.AttackCooldownMax.HasValue)
            enemy.AttackCooldownMax *= (float)style.AttackCooldown;
        if (enemy.AttackCooldown.HasValue)
            enemy.AttackCooldown *= (float)style.AttackCooldown;
        enemy.ScaleAttackRange(style.AttackRange);
        enemy.AwarenessRange *= (float)style.Awareness;
        enemy.DisengageRange *= (float)style.Awareness;
        enemy.InteractionTags = enemy.InteractionTags.Concat(style.Tags ?? Array.Empty<string>()).ToHashSet();
        foreach (var child in enemy.SpawnedEnemies)
            ApplyEnemyIdentity(child);
        return enemy;
    }

    /// <summary>Applies active-path projectile rules to newly emitted hostile shots.</summary>
    public static void TuneNewProjectiles(IEnumerable<EnemyProjectile> projectiles)
    {
        var path = Active();
        var style = path.Style;
        if (path.Key == "sound")
            return;

        foreach (var projectile in projectiles)
        {
            if (projectile.ContentPath == path.Key)
                continue;
            projectile.ContentPath = path.Key;
            projectile.Speed *= (float)style.ProjectileSpeed;
            projectile.Size *= (float)style.ProjectileSize;
            projectile.Damage = MathF.Round(projectile.Damage * (float)style.ProjectileDamage);
            projectile.RemainingRange *= (float)style.ProjectileRange;
            if (style.ProjectileLifetime.HasValue)
                projectile.Lifetime = Math.Max(projectile.Lifetime ?? 0, (float)style.ProjectileLifetime.Value);
            if (style.AimedChance < 1 && ProjectileRng.NextDouble() > style.AimedChance)
                projectile.Direction += (float)(ProjectileRng.NextDouble() * Math.PI * 2 - Math.PI);
            if (style.Colors is { Count: > 0 })
                projectile.Color = style.Colors[^1];
        }
    }
}
